### PR TITLE
Separate tweaking from key aggregation

### LINKED ISSUE #12: Rework KeyAgg algorithm to simplify Tweaking
The following issue was discussed on the [mailing list](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-April/020372.html): To apply a tweaking scheme like Taproot, access to the internal key is required. Hence, one needs to call KeyAgg multiple times:
```
P := KeyAgg(pk1..u, <empty>, <empty>):
tweak := hash_TapTweak(P, root)
is_xonly_t := true
Q := KeyAgg(pk1..u, tweak, is_xonly_t):
```

Not only is this inefficient but it's also unintuitive to the point that no implementation that allows tweaking will implement KeyAgg as described in the spec. This is a good reason for changing the spec.

The most reasonable approach to me at this point is that we split the existing `KeyAgg` into a `KeyAgg` algoritm that only does key aggregation, `ApplyTweak` and `ApplyXonlyTweak` - similar to the libsecp-zkp API. The downside is that we need to expose the key aggregation state that was previously fully contained inside `KeyAgg`.

### PR DESCRIPTION
**Pull request by jonasnick**
Fixes #12. This significantly simplifies BIP32 and Taproot tweaking because the public key that is supposed to be tweaked is easily available. Based on #9. 

I think this definition of ApplyTweak and KeyAgg Context can be reused for a bip-FROST.

### CODE REVIEW

File: bip-musig2.mediawiki [hunk 1]
```diff
@@ -135,13 +138,17 @@ However, if ''PartialSigVerify'' succeeds for all partial signatures then ''Part
...
+
 In addition to public keys, the ''KeyAgg'' algorithm accepts tweaks, which modify the aggregate public key as defined in the [[#tweaking-definition|Tweaking Definition]] subsection.
 For example, if ''KeyAgg'' is run with ''v = 2'', ''is_xonly_t<sub>1</sub> = false'', ''is_xonly_t<sub>2</sub> = true'', then the aggregate key is first ordinarily tweaked with ''tweak<sub>1</sub>'' and then X-only tweaked with ''tweak<sub>2</sub>''.
 
 The purpose of specifying tweaking is to ensure compatibility with existing uses of tweaking, i.e., that the result of signing is a valid signature for the tweaked public key.
 The MuSig2 algorithms take arbitrary tweaks as input but accepting arbitrary tweaks may negatively affect the protocol's security.
 Instead, signers should obtain the tweaks according to other specifications.
-This typically involves deriving the tweaks from a hash of the aggregate public key and some other information.
+This typically involves deriving the tweaks from a hash of the aggregate public key (obtained with ''GetPubkey'' or ''GetLegacyPubkey'') and some other information.
```
Comment (robot-dreams) [inline bip-musig2.mediawiki]: Instead of introducing the "Legacy" terminology here (that could result in potential confusion), how would you feel about `GetPubkeyBytes` and `GetPubkeyCbytes`?

File: bip-musig2.mediawiki [hunk 2]
```diff
@@ -282,7 +307,7 @@ Input:
 * If ''len(sk) > 0'':
 ** Let ''rand'' be the byte-wise xor of ''sk'' and ''hash<sub>MuSig/aux</sub>(rand')''<ref>The random data is hashed (with a unique tag) as a precaution against situations where the randomness may be correlated with the secret signing key itself. It is xored with the secret key (rather than combined with it in a hash) to reduce the number of operations exposed to the actual secret key.</ref>.
 * Else: let ''rand = rand' ''
-* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(1, i) || bytes(1, len(m)) || m || bytes(4, len(extra_in)) || extra_in)) mod n'' for ''i = 1,2''
+* Let ''k<sub>i</sub> = int(hash<sub>MuSig/nonce</sub>(rand || bytes(1, len(aggpk)) || aggpk || bytes(1, len(m)) || m || bytes(4, len(extra_in)) || extra_in || bytes(1, i))) mod n'' for ''i = 1,2''
```
Comment (robot-dreams) [inline bip-musig2.mediawiki]: I'm assuming some commits are repeated between this PR and #9 , and that any minor inconsistencies (e.g. in this PR the spec includes this nonce_hash update but the Python doesn't) will be resolved after merging #9 and rebasing.

Comment (robot-dreams) [inline bip-musig2.mediawiki]: Should this change be isolated to a separate PR (especially since the corresponding Python change isn't in the current PR)?

File: bip-musig2.mediawiki [hunk 3]
```diff
@@ -430,27 +457,27 @@ In order to produce a partial signature for an X-only public key that is an aggr
 
 <poem>
 The following elliptic curve points arise as intermediate steps in the MuSig2 protocol:
-• ''P<sub>i</sub>'' as computed in ''KeyAggInternal'' is the point corresponding to the ''i''-th signer's X-only public key. Defining ''d'<sub>i</sub>'' to be the ''i''-th signer's secret key as an integer, i.e. the ''d' '' value as computed in the ''Sign'' algorithm of the ''i''-th signer, we have
+• ''P<sub>i</sub>'' as computed in ''KeyAgg'' is the point corresponding to the ''i''-th signer's X-only public key. Defining ''d'<sub>i</sub>'' to be the ''i''-th signer's secret key as an integer, i.e. the ''d' '' value as computed in the ''Sign'' algorithm of the ''i''-th signer, we have
     ''P<sub>i</sub> = with_even_y(d'<sub>i</sub>⋅G) ''.
-• ''Q<sub>0</sub>'' is an aggregate of the signers' public keys and defined in ''KeyAggInternal''  as
+• ''Q<sub>0</sub>'' is the aggregate of the signers' public keys and defined in ''KeyAgg''  as
```
Comment (robot-dreams) [inline bip-musig2.mediawiki]: Nit: `KeyAgg` now defines `Q` rather than `Q<sub>0</sub>`, right?

Comment (robot-dreams) [inline bip-musig2.mediawiki]: Nit: `KeyAgg` now defines `Q` rather than `Q<sub>0</sub>`, right?

File: bip-musig2.mediawiki [hunk 4]
```diff
@@ -430,27 +457,27 @@ In order to produce a partial signature for an X-only public key that is an aggr
...
     ''Q<sub>0</sub> = a<sub>1</sub>⋅P<sub>1</sub> + a<sub>2</sub>⋅P<sub>2</sub> + ... + a<sub>u</sub>⋅P<sub>u</sub>''.
-• ''Q<sub>i</sub>'' as computed in ''ApplyTweak'' for ''1 &le; i &le; v'' is the tweaked public key after the ''i''-th tweaking operation. It holds that
+• ''Q<sub>i</sub>'' is the tweaked public key after the ''i''-th execution of ''ApplyTweak'' for ''1 &le; i &le; v''. It holds that
     ''Q<sub>i</sub> = f(i-1) + t<sub>i</sub>⋅G'' for ''i = 1, ..., v'' where
         ''f(i-1) := with_even_y(Q<sub>i-1</sub>)'' if ''is_xonly_t<sub>i</sub>'' and
         ''f(i-1) := Q<sub>i-1</sub>'' otherwise.
-• ''with_even_y(Q<sub>v</sub>)'' is the final result of ''KeyAgg''.
+• ''with_even_y(Q<sub>v</sub>)'' is the final result of the key aggregation and tweaking operations. It corresponds to the output of ''GetPubkey'' applied on they final KeyAgg Context.
```
Comment (robot-dreams) [inline bip-musig2.mediawiki]: Nit: `on they` -> `on the`

Comment (robot-dreams) [inline bip-musig2.mediawiki]: Nit: `on they` -> `on the`

**Review by robot-dreams — COMMENTED** [review]
I'm a fan of this approach! Overall looks good; I compared the spec and code updates side by side, and they look compatible.

I didn't check test vectors carefully because there seems to be some overlap with #9.

Given there are a lot of in-flight changes I think it's worth being careful about merge order / conflicts.

File: bip-musig2.mediawiki (hunk 1 above)
Comment (jonasnick) [reply bip-musig2.mediawiki]: This solution isn't ideal either because the definition of a pubkey is already that it's a byte array. How about replacing `GetLegacyPubkey` with `GetPubkey33` to avoid the word "compressed" because it's not compressed compared to the (x-only) pubkey?

File: bip-musig2.mediawiki (hunk 2 above)
Comment (jonasnick) [reply bip-musig2.mediawiki]: This is only in this PR because I based this PR on another PR. This particular change has already gone away because the corresponding PR got merged and this PR was rebased.

File: bip-musig2.mediawiki (hunk 3 above)
Comment (jonasnick) [reply bip-musig2.mediawiki]: Fixed.

File: bip-musig2.mediawiki (hunk 4 above)
Comment (jonasnick) [reply bip-musig2.mediawiki]: Fixed

**Comment by robot-dreams** [comment]
ACK 88c25a99bf581e743ef5450ded61f2c7bb6591fe

**Comment by jonasnick** [comment]
rebased

File: bip-musig2.mediawiki [hunk 5]
```diff
@@ -440,27 +467,27 @@ In order to produce a partial signature for an X-only public key that is an aggr
 
 <poem>
 The following elliptic curve points arise as intermediate steps in the MuSig2 protocol:
-• ''P<sub>i</sub>'' as computed in ''KeyAggInternal'' is the point corresponding to the ''i''-th signer's X-only public key. Defining ''d'<sub>i</sub>'' to be the ''i''-th signer's secret key as an integer, i.e. the ''d' '' value as computed in the ''Sign'' algorithm of the ''i''-th signer, we have
+• ''P<sub>i</sub>'' as computed in ''KeyAgg'' is the point corresponding to the ''i''-th signer's X-only public key. Defining ''d'<sub>i</sub>'' to be the ''i''-th signer's secret key as an integer, i.e. the ''d' '' value as computed in the ''Sign'' algorithm of the ''i''-th signer, we have
     ''P<sub>i</sub> = with_even_y(d'<sub>i</sub>⋅G) ''.
-• ''Q<sub>0</sub>'' is an aggregate of the signers' public keys and defined in ''KeyAggInternal''  as
+• ''Q<sub>0</sub>'' is the aggregate of the signers' public keys and defined as ''Q'' in ''KeyAgg'':
```
Comment (real-or-random) [inline bip-musig2.mediawiki]:
```suggestion
• ''Q<sub>0</sub>'' is the aggregate of the signers' public keys and defined like ''Q'' in ''KeyAgg'':
```

Not sure if this is grammatical but it took me a while to parse the current sentence because I expect a definition after "is defined as"... We could also simply rename Q to Q0 in KeyAgg. (I mean, KeyAgg now really outputs the first key, so Q0 makes sense.)

File: bip-musig2.mediawiki [hunk 6]
```diff
@@ -268,16 +285,24 @@ Input:
...
+Input:
+* The ''keygen_ctx'': a [[#keygen-context|KeyGen Context]] data structure
+* The ''tweak'': a 32-byte array
+* The tweak mode ''is_xonly_t'': a boolean
+
+'''''ApplyTweak(keygen_ctx, tweak, is_xonly_t)''''':
+* Let ''(Q, gacc, tacc) = keygen_ctx''
+* If ''is_xonly_t'' and ''not has_even_y(Q)'':
+** Let ''g = -1 mod n''
+* Else: let ''g = 1''
```
Comment (real-or-random) [inline bip-musig2.mediawiki]:
```suggestion
* Else:
** Let ''g = 1''
```

File: bip-musig2.mediawiki [hunk 7]
```diff
@@ -145,13 +148,17 @@ However, if ''PartialSigVerify'' succeeds for all partial signatures then ''Part
 
 ==== Tweaking ====
 
+The aggregate public key can be ''tweaked'', which modifies the key as defined in the [[#tweaking-definition|Tweaking Definition]] subsection.
+In order to apply a tweak, the KeyGen Context output by ''KeyAgg'' is provided to the ''ApplyTweak'' algorithm with the ''is_xonly_t'' argument set to false for ordinary tweaking and true for X-only tweaking.
+The resulting KeyGen Context can be used to apply another tweak with ''ApplyTweak'' or obtain the aggregate public key with ''GetPubkey''.
```
Comment (real-or-random) [inline bip-musig2.mediawiki:153]: Should this mention GetPubkey33 already? I don't think it's really necessary to mention it here but...

File: bip-musig2.mediawiki [hunk 8]
```diff
@@ -145,13 +148,17 @@ However, if ''PartialSigVerify'' succeeds for all partial signatures then ''Part
...
+
 In addition to public keys, the ''KeyAgg'' algorithm accepts tweaks, which modify the aggregate public key as defined in the [[#tweaking-definition|Tweaking Definition]] subsection.
 For example, if ''KeyAgg'' is run with ''v = 2'', ''is_xonly_t<sub>1</sub> = false'', ''is_xonly_t<sub>2</sub> = true'', then the aggregate key is first ordinarily tweaked with ''tweak<sub>1</sub>'' and then X-only tweaked with ''tweak<sub>2</sub>''.
 
 The purpose of specifying tweaking is to ensure compatibility with existing uses of tweaking, i.e., that the result of signing is a valid signature for the tweaked public key.
 The MuSig2 algorithms take arbitrary tweaks as input but accepting arbitrary tweaks may negatively affect the protocol's security.
 Instead, signers should obtain the tweaks according to other specifications.
-This typically involves deriving the tweaks from a hash of the aggregate public key and some other information.
+This typically involves deriving the tweaks from a hash of the aggregate public key (obtained with ''GetPubkey'' or ''GetPubkey33'') and some other information.
```
Comment (real-or-random) [inline bip-musig2.mediawiki]: ... here I was surprised that it appears of the blue.

File: bip-musig2.mediawiki (hunk 1 above)
Comment (real-or-random) [reply bip-musig2.mediawiki]: Hm, I think `GetPubkey33` is better but I still find it somewhat confusing because for tweaking we different terms "ordinary" and "xonly". But I see that a "pubkey" here is by default an xonly pubkey, so that doesn't really make sense either. :/

Sorry this is not constructive. I don't have a better suggestion than we have now but if you have other ideas, it may be worth mentioning them.

**Review by real-or-random — COMMENTED** [review]
ACK mod nits, nice to get rid of some of the subscripts

File: bip-musig2.mediawiki (hunk 7 above)
Comment (jonasnick) [reply bip-musig2.mediawiki:153]: Hm, here I wanted to only mention how to get the public key for Schnorr verification after tweaking.

File: bip-musig2.mediawiki (hunk 8 above)
Comment (jonasnick) [reply bip-musig2.mediawiki]: Added a few sentences to explain `GetPubkey33`

File: bip-musig2.mediawiki (hunk 5 above)
Comment (jonasnick) [reply bip-musig2.mediawiki]: It seems slightly confusing to have `Q<sub>0</sub>` in the algorithm description but no `Q` with another subscript. Tried improving the wording.

File: bip-musig2.mediawiki (hunk 6 above)
Comment (jonasnick) [reply bip-musig2.mediawiki]: fixed

Comment (real-or-random) [reply bip-musig2.mediawiki]: You said "fixed" here but you didn't update the branch.

Comment (jonasnick) [reply bip-musig2.mediawiki]: Sorry, should be done now.

**Comment by jonasnick** [comment]
Squashed fixup commit

**Comment by real-or-random** [comment]
needs rebase

**Comment by jonasnick** [comment]
done

**Review by real-or-random — COMMENTED** [review]
```
❯ mypy bip-musig2/reference.py
bip-musig2/reference.py:300: error: Argument 1 to "bytes_from_point" has incompatible type "Optional[Tuple[int, int]]"; expected "Tuple[int, int]"
bip-musig2/reference.py:316: error: Argument 1 to "has_even_y" has incompatible type "Optional[Tuple[int, int]]"; expected "Tuple[int, int]"
bip-musig2/reference.py:317: error: Argument 1 to "has_even_y" has incompatible type "Optional[Tuple[int, int]]"; expected "Tuple[int, int]"
bip-musig2/reference.py:331: error: Argument 1 to "cbytes" has incompatible type "Optional[Tuple[int, int]]"; expected "Tuple[int, int]"
bip-musig2/reference.py:349: error: Argument 1 to "has_even_y" has incompatible type "Optional[Tuple[int, int]]"; expected "Tuple[int, int]"
bip-musig2/reference.py:369: error: Argument 1 to "bytes_from_point" has incompatible type "Optional[Tuple[int, int]]"; expected "Tuple[int, int]"
Found 6 errors in 1 file (checked 1 source file)
```
ACK apart from that

**Comment by jonasnick** [comment]
rebase fixed the mypy errors

**Review by real-or-random — APPROVED** [review]
ACK
